fv3config/filesystem.py

```diff
@@ -0,0 +1,59 @@
+import os
+import shutil
+from ._exceptions import DelayedImportError
+from fsspec.implementations.local import LocalFileSystem
+try:
+    import gcsfs
+except ImportError as err:
+    gcsfs = DelayedImportError(err)
+try:
+    import google.auth
+except ImportError as err:
+    google = DelayedImportError(err)
+
+LOCAL_FS = LocalFileSystem()
+GS_BUCKET_PREFIX = 'gs://'
+
+
+def _put_directory(local_source_dir, dest_dir, fs=None):
+    """Copy the contents of a local directory to a local or remote directory.
+    """
+    if fs is None:
+        fs = _get_fs(dest_dir)
+    for token in os.listdir(local_source_dir):
+        source = os.path.join(os.path.abspath(local_source_dir), token)
+        dest = os.path.join(dest_dir, token)
+        if os.path.isdir(source):
+            fs.makedirs(dest, exist_ok=True)
+            _put_directory(source, dest, fs)
+        else:
+            _put_file(source, dest, fs)
+
+
+def _get_file(source_filename, dest_filename, fs=None):
+    fs = _get_fs(source_filename)
+    fs.get(source_filename, dest_filename)
+
+
+def _put_file(source_filename, dest_filename, fs=None):
+    fs = _get_fs(dest_filename)
+    fs.put(source_filename, dest_filename)
+
+
+def _is_gcloud_path(path):
+    return path.startswith(GS_BUCKET_PREFIX)
+
+
+def _get_gcloud_project():
+    return os.environ.get(google.auth.environment_vars.PROJECT)
+
+
+def _get_fs(path):
+    if _is_gcloud_path(path):
```

Comment:
I think that would be appropriate when we decide we want to support more than just gcs as a remote file system. As it stands, the current code is easier to read than this example and does the same thing in half the number of lines. But certainly it would be cleaner to call fsspec.filesystem(proto) when proto can take on more than one possible value (currently only 'gs').

Supporting alternate remote filesystems will be significantly more complicated than just accessing additional fsspec options - authentication has been by far most of the work in getting gcsfs to work.

Reply:
I would say though in the long run that's probably the way we want to go, and we can ensure it will be easy to switch over to that usage by always instantiating the fs object using `_get_fs()`.

Reply:
I don't think it's harder to read, and it would allow you to remove a significant chunk of the code in this module. It would allow you to delete `_is_gcloud_path`, `_get_gcloud_fs`, `_get_gcloud_project`, as well as the manual import and instantiation `LocalFileSystem`. I see no reason not to use the more general purpose solution in this case.

Reply:
I think this module is really a good demonstration of why there are design patterns (e.g. the factory pattern, prototype pattern, etc) for creating objects.

Reply:
I've combined what you're saying about being able to delete _is_gcloud_path etc. with what I was saying about writing the code for only `gs://` at the moment. Everything is now completely agnostic to the type of remote filesystem being used, so adding remote filesystems will really be as simple as modifying `get_fs`.

I also felt `get_fs` is a central enough method that I removed the underscore, even if it's meant to be used internally only and probably shouldn't be imported into the top-level namespace. Its API should remain stable.